### DISCUSSION TITLE
Using `pnpm` as the npm package manager

### DISCUSSION THREAD

**Discussion by erha19** [opensumi/core#1572] [RFC]
当前 `pnpm` 在运行效率及空间上具有一定优势（相较于 npm、yarn），介绍见 https://github.com/pnpm/pnpm

部分社区一线项目也逐渐切换至了 `pnpm`, 如 [vite](https://github.com/vitejs/vite), [vue](https://github.com/vuejs/core) 等。

将 OpenSumi 的包管理器也替换为 `pnpm` 可能遇到的问题：

1.  项目中可能存在部分循环/幽灵依赖需要处理
2. 部分构建逻辑可能会受影响，如 native rebuild 相关逻辑

相关 Issue：

- https://github.com/opensumi/ide-electron/pull/43

欢迎下面留下你的意见以及投票 ~

**Comment by EAGzzyCSL** [comment]
循环依赖和幽灵依赖 eslint-plugin-import 应该可以锤出来，剩下就是要人肉 fix 了

- 循环依赖：https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-cycle.md
- 幽灵依赖：https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/no-extraneous-dependencies.md

**Comment by yantze** [comment]
看起来 pnpm 支持 resolutions，那应该可以平滑迁移过去

**Comment by BlackHole1** [comment]
基于 @EAGzzyCSL 的评论，我们可以得出以下结论：
1. 循环依赖/幽灵依赖 可以通过 `eslint-plugin-import` 来找出问题代码，并进行人工 fix
2. **循环依赖/幽灵依赖** 也可以使项目本身具备更好的维护性
3. 解决 **循环依赖** 有助于未来的 webpack -> vite (如: ide-electron 项目)
   * electron 使用 vite 时，开发体验将大幅度提高

同时 pnpm 具备了 `shamefully-hoist` 功能，开启后可以让 pnpm 对 node_modules 目录处理时，更接近 yarn，可以在一定程度上使用此功能，让项目快速迁移到 pnpm。

**Comment by BlackHole1** [comment]
有什么进展么？通过现阶段的讨论，感觉都是同意迁移到 pnpm。 

/cc @erha19

**Comment by hacke2** [comment]
+1

**Comment by erha19** [comment]
@BlackHole1 我们会在 2.21.0 迭代进行这个事情 ~ 如果有兴趣的话也可以欢迎直接往我们项目提代码

**Reply by BlackHole1** [reply]
好的。
那 https://github.com/opensumi/ide-electron/pull/43 此 PR 是否可以进行合并了？
